src/blas/backends/rocblas/rocblas_batch.cpp

```diff
@@ -732,344 +976,505 @@ TRSM_BATCH_LAUNCHER_USM(std::complex<double>, rocblas_ztrsm_batched)
-                sycl::buffer<std::complex<double>, 1> &x, int64_t incx, int64_t stridex,
-                sycl::buffer<std::complex<double>, 1> &y, int64_t incy, int64_t stridey,
-                int64_t batch_size) {
-    throw unimplemented("blas", "axpy_batch", "for row_major layout");
-}
+#undef AXPY_STRIDED_BATCH_LAUNCHER
 
-void gemv_batch(sycl::queue &queue, transpose transa, int64_t m, int64_t n, float alpha,
-                sycl::buffer<float, 1> &a, int64_t lda, int64_t stride_a, sycl::buffer<float, 1> &x,
-                int64_t incx, int64_t stride_x, float beta, sycl::buffer<float, 1> &y, int64_t incy,
-                int64_t stride_y, int64_t batch_size) {
-    throw unimplemented("blas", "gemv_batch", "for row_major layout");
-}
+template <typename Func, typename T>
+inline void gemv_batch(Func func, sycl::queue &queue, transpose trans, int64_t m, int64_t n,
+                       std::complex<T> alpha, sycl::buffer<std::complex<T>, 1> &a, int64_t lda,
+                       int64_t stridea, sycl::buffer<std::complex<T>, 1> &x, int64_t incx,
+                       int64_t stridex, std::complex<T> beta, sycl::buffer<std::complex<T>, 1> &y,
+                       int64_t incy, int64_t stridey, int64_t batch_size) {
+    auto new_trans = trans == oneapi::mkl::transpose::nontrans ? oneapi::mkl::transpose::trans
+                                                               : oneapi::mkl::transpose::nontrans;
+
+    if (trans == oneapi::mkl::transpose::conjtrans) {
+        alpha = std::conj(alpha);
+        beta = std::conj(beta);
+
+        auto tmp_x = incx;
+        auto tmp_y = incy;
+        auto tmp_stridex = stridex;
+        auto tmp_stridey = stridey;
+
+        incx = std::abs(incx);
+        incy = std::abs(incy);
+        stridex = std::abs(stridex);
+        stridey = std::abs(stridey);
+
+        if (m > 0) {
+            queue.submit([&](sycl::handler &cgh) {
+                auto acc = x.template get_access<sycl::access::mode::read_write>(cgh);
+                cgh.parallel_for(sycl::range{ (std::size_t)batch_size, (std::size_t)m },
+                                 [=](sycl::item<2> it) {
+                                     auto index = it.get_id(0) * stridex + it.get_id(1) * incx;
+                                     acc[index] = std::conj(acc[index]);
+                                 });
+            });
+
+            if (n > 0) {
+                queue.submit([&](sycl::handler &cgh) {
+                    auto acc = y.template get_access<sycl::access::mode::read_write>(cgh);
+                    cgh.parallel_for(sycl::range{ (std::size_t)batch_size, (std::size_t)n },
+                                     [=](sycl::item<2> it) {
+                                         auto index = it.get_id(0) * stridey + it.get_id(1) * incy;
+                                         acc[index] = std::conj(acc[index]);
+                                     });
+                });
+            }
+        }
```

Comment:
# Can you explain what is the purpose and need here?

The reason needed for `x` to be conjugated is that we are dealing in this case with row-major layout.

Emulating the transpose operation can be done by switching the "propriety" arguments of the vectors/matrices, thus modifying the way the inputs are accessed as opposed to physically transposing the inputs.

However for complex conjugations transpose, for elements in vector/matrices, this operation must be done
explicitly since there is no way to "emulate" the effect of the conjugate transpose operation.

# Why do we need to take the conjugate of vectors exclusively?

From my readings during the work of this PR, I realized that for the TRSM function, as an example, which does not need to have the conjugate emulation, is because triangular matrices do not change upon transposition.

Proprieties like the one mentioned for TRSM are the reasons manual conjugation is needed or not needed.

Also, the source of truth for the oneMKL's tests is the CBLAS's lapack reference implementation, and this is how they implemented them

Reply:
Thanks for the explanation. Makes sense.